devlib/utils/ssh.py

```diff
@@ -170,51 +305,422 @@ def name(self):
+            # file-like object
+            else:
+                r = stream_out
+                w = stream_out
+
+            return (r, w)
+
+        out_streams = {
+            name: create_out_stream(stream_in, stream_out)
+            for stream_in, stream_out, name in (
+                (stdout_in, stdout, 'stdout'),
+                (stderr_in, stderr, 'stderr'),
+            )
+        }
+
+        def redirect_thread_f(stdout_in, stderr_in, out_streams, select_timeout):
+            def callback(out_streams, name, chunk):
+                try:
+                    r, w = out_streams[name]
+                except KeyError:
+                    return out_streams
+
+                try:
+                    w.write(chunk)
+                # Write failed
+                except ValueError:
+                    # Since that stream is now closed, stop trying to write to it
+                    del out_streams[name]
+                    # If that was the last open stream, we raise an
+                    # exception so the thread can terminate.
+                    if not out_streams:
+                        raise
+
+                return out_streams
+
+            try:
+                _read_paramiko_streams(stdout_in, stderr_in, select_timeout, callback, copy.copy(out_streams))
+            # The streams closed while we were writing to it, the job is done here
+            except ValueError:
+                pass
+
+            # Make sure the writing end are closed proper since we are not
+            # going to write anything anymore
+            for r, w in out_streams.values():
+                if r is not w and w is not None:
+                    w.close()
+
+        # If there is anything we need to redirect to, spawn a thread taking
+        # care of that
+        select_timeout = 1
+        thread_out_streams = {
+            name: (r, w)
+            for name, (r, w) in out_streams.items()
+            if w is not None
+        }
+        redirect_thread = threading.Thread(
+            target=redirect_thread_f,
+            args=(stdout_in, stderr_in, thread_out_streams, select_timeout),
+            # The thread will die when the main thread dies
+            daemon=True,
+        )
+        redirect_thread.start()
+
+        return ParamikoBackgroundCommand(
+            conn=self,
+            as_root=as_root,
+            chan=channel,
+            pid=pid,
+            stdin=stdin,
+            # We give the reading end to the consumer of the data
+            stdout=out_streams['stdout'][0],
+            stderr=out_streams['stderr'][0],
+            redirect_thread=redirect_thread,
+        )
+
+    def _close(self):
+        logger.debug('Logging out {}@{}'.format(self.username, self.host))
+        with _handle_paramiko_exceptions():
+            bg_cmds = set(self._current_bg_cmds)
+            for bg_cmd in bg_cmds:
+                bg_cmd.close()
+            self.client.close()
```

Comment:
Good point, I used `__del__` and I guess that's one more occurence of someone being bitten by it. The exception should be harmless though as it's swallowed by the interpreter when coming from `__del__`. Maybe we should hide `AttributeError` explicitly to avoid ugly messages being printed.

Reply:
I'm not sure sure about just hiding the error as this may lead to confusion in the future. Would it be worth declaring the attribute explicitly as None before initialising the connection and checking if the client exists before attempting to close it?

Reply:
If we want to avoid calling `close()` until `__init__` is finished, we could add a checkpoint at the end of `__init__` with a metaclass quite easily, so `__del__` would now what is safe to do. This would work well with subclassing, and would not be susceptible to mistakes in the future when someone modifies the classes.